sktime/transformations/base.py

```diff
@@ -48,93 +104,537 @@ class BaseTransformer(BaseEstimator):
+            if y is not None:
+                ValueError(
+                    "no default behaviour if _fit does not support Panel, "
+                    " but X is Panel and y is not None"
+                )
+            X = convert_to(X, to_type="df-list", as_scitype="Panel")
+
+            if self.get_tag("fit-in-transform"):
+                Xt = [clone(self).transform(Xi) for Xi in X]
+            else:
+                transformers = self.transformers_
+                if len(transformers) != len(X):
+                    raise RuntimeError(
+                        "found different number of instances in transform than in fit"
+                    )
+                else:
+                    Xt = [transformers[i].transform(X[i]) for i in range(len(X))]
+            # now we have a list of transformed instances
+
+            # if the output is Series, Xt is a Panel and we convert back
+            if output_scitype == "Series":
+                Xt = convert_to(X, to_type=X_input_mtype, as_scitype="Panel")
+            # if the output is Primitives, we have a list of one-row dataframes
+            # we concatenate those and overwrite the index with that of X
+            elif output_scitype == "Primitives":
+                Xt = pd.concat(Xt)
+                Xt.index = X.index
+            return Xt
+
+        # convert X/y to supported inner type, if necessary
+        ###################################################
+
+        # variables for the scitype of the current X (possibly converted)
+        #     y wasn't converted so we can use y_input_scitype
+        X_mtype = mtype(X)
+        X_scitype = mtype_to_scitype(X_mtype)
+
+        # subset to the mtypes that are of the same scitype as X/y
+        X_inner_mtype = [
+            mt for mt in X_inner_mtype if mtype_to_scitype(mt) == X_scitype
+        ]
+
+        y_inner_mtype = [
+            mt for mt in y_inner_mtype if mtype_to_scitype(mt) == y_input_scitype
+        ]
+
+        # convert X and y to a supported internal type
+        #  if X/y type is already supported, no conversion takes place
+        X_inner = convert_to(
+            X,
+            to_type=X_inner_mtype,
+            as_scitype=X_scitype,
+        )
+        y_inner = convert_to(
+            y,
+            to_type=y_inner_mtype,
+            as_scitype=y_input_scitype,
+        )
+
+        # carry out the transformation
+        ###################################################
+
+        # uncomment this once Z is completely gone
+        # Xt = self._transform(X=X_inner, y=y_inner)
+        # less robust workaround until then
+        Xt = self._transform(X_inner, y_inner)
+
+        # convert transformed X back to input mtype
+        ###########################################
+
+        # if we converted Series to "one-instance-Panel", revert that
+        if X_was_Series:
+            Xt = convert_Panel_to_Series(Xt)
```

Comment:
@fkiraly In the case where `transform-output` is Primitive, I think `Xt` will have type pd.DataFrame, not panel. I think this would break this function

Reply:
yes, that is a bug - fixed